#### Importing pandas

In [1]:
import pandas as pd
import numpy as np

#### Importing the data

In [2]:
geo=pd.read_csv(r"/content/cells_geo.csv")
day1=pd.read_csv(r"/content/Telcom_dataset.csv")
day2=pd.read_csv(r"/content/Telcom_dataset2.csv")
day3=pd.read_csv(r"/content/Telcom_dataset3.csv")
geo_description=pd.read_excel(r"/content/cells_geo_description.xlsx")
CDR_description=pd.read_excel(r"/content/CDR_description.xlsx")

#### Exploring the datasets

##### 1. Geo

In [3]:
geo.head()

;VILLES;STATUS;LOCALISATION;DECOUPZONE;ZONENAME;LONGITUDE;LATITUDE;REGION;AREA;CELL_ID;SITE_CODE
0  0;ADJAME;In Service;ABIDJAN;"""Abidjan_EST";AG...                                              
1  1;ADJAME;In Service;ABIDJAN;"""Abidjan_EST";AG...                                              
2  2;ADJAME;In Service;ABIDJAN;"""Abidjan_EST";AG...                                              
3  3;ASSINIE;In Service;INTERIEUR;Grand-EST;"ASSI...                                              
4  4;ASSINIE;In Service;INTERIEUR;Grand-EST;"ASSI...

In [4]:
geo.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3974 entries, 0 to 3973
Data columns (total 1 columns):
 #   Column                                                                                            Non-Null Count  Dtype 
---  ------                                                                                            --------------  ----- 
 0   ;VILLES;STATUS;LOCALISATION;DECOUPZONE;ZONENAME;LONGITUDE;LATITUDE;REGION;AREA;CELL_ID;SITE_CODE  3974 non-null   object
dtypes: object(1)
memory usage: 31.2+ KB


**Separating the columns**

In [5]:
geo=geo[";VILLES;STATUS;LOCALISATION;DECOUPZONE;ZONENAME;LONGITUDE;LATITUDE;REGION;AREA;CELL_ID;SITE_CODE"].str.split(';',
                                                                                                                      expand=True)

In [6]:
geo.head()

0        1           2          3   ...         8  9           10          11
0  0   ADJAME  In Service    ABIDJAN  ...     LAGUNE     fbb2e75af3  0bb492ef17
1  1   ADJAME  In Service    ABIDJAN  ...     LAGUNE     1502501d78  0bb492ef17
2  2   ADJAME  In Service    ABIDJAN  ...     LAGUNE     0c2e182a0c  0bb492ef17
3  3  ASSINIE  In Service  INTERIEUR  ...  SUD COMOE     69bdfc69fd  8a1e6ce31f
4  4  ASSINIE  In Service  INTERIEUR  ...  SUD COMOE     7a1029d2ab  8a1e6ce31f

[5 rows x 12 columns]

**Renaming the columns correctly**

In [7]:
geo=geo.rename(columns={1:"VILLES",
                    2:"STATUS",
                    3:"LOCALISATION",
                    4:"DECOUPZONE",
                    5:"ZONENAME",
                    6:"LONGITUDE",
                    7:"LATITUDE",
                    8:"REGION",
                    9:"AREA",
                    10:"CELL_ID",
                    11:"SITE_CODE"
                  })

In [8]:
geo.drop(0,inplace=True, axis=1)

In [9]:
geo.head()

VILLES      STATUS LOCALISATION  ... AREA     CELL_ID   SITE_CODE
0   ADJAME  In Service      ABIDJAN  ...       fbb2e75af3  0bb492ef17
1   ADJAME  In Service      ABIDJAN  ...       1502501d78  0bb492ef17
2   ADJAME  In Service      ABIDJAN  ...       0c2e182a0c  0bb492ef17
3  ASSINIE  In Service    INTERIEUR  ...       69bdfc69fd  8a1e6ce31f
4  ASSINIE  In Service    INTERIEUR  ...       7a1029d2ab  8a1e6ce31f

[5 rows x 11 columns]

**Summary of the Geo data**

In [10]:
geo_description

Column name                                        Description  Format
0         VILLES                                               City  String
1         STATUS                                  In Service or not  String
2   LOCALISATION                                  in ABIDJAN or not  String
3     DECOUPZONE                                  Geographical Zone  String
4       ZONENAME                                       Name of Zone  String
5      LONGITUDE                                          Longitude   Float
6       LATITUDE                                           Latitude   Float
7         REGION                                             Region  String
8           AREA                                               Area  String
9        CELL_ID                                     ID of the cell  String
10     SITE_CODE  Site (there are several cells per site, severa...  String

In [11]:
geo.describe(include="all")

VILLES      STATUS LOCALISATION  ...           AREA     CELL_ID   SITE_CODE
count     3974        3974         3974  ...           3974        3974        3974
unique     504           2            2  ...              7        3916        1029
top     COCODY  In Service    INTERIEUR  ...  ABIDJAN NORTH  6750e0fb7b  20387d4292
freq       354        3907         2305  ...           1222          39          42

[4 rows x 11 columns]

In [12]:
geo.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3974 entries, 0 to 3973
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   VILLES        3974 non-null   object
 1   STATUS        3974 non-null   object
 2   LOCALISATION  3974 non-null   object
 3   DECOUPZONE    3974 non-null   object
 4   ZONENAME      3974 non-null   object
 5   LONGITUDE     3974 non-null   object
 6   LATITUDE      3974 non-null   object
 7   REGION        3974 non-null   object
 8   AREA          3974 non-null   object
 9   CELL_ID       3974 non-null   object
 10  SITE_CODE     3974 non-null   object
dtypes: object(11)
memory usage: 341.6+ KB


- There are no null values because every row is in string format

In [13]:
geo["DECOUPZONE"].unique()

array(['"""Abidjan_EST"', 'Grand-EST', 'Grand-OUEST', '"""Abidjan_NORD"',
       '"""Abidjan_CENTRE"', 'Abidjan_NORD', 'Abidjan_EST',
       'Abidjan_OUEST', 'Grand-SUD', 'SUD-OUEST', 'Abidjan_CENTRE',
       'Abidjan_SUD', 'Grand-NORD', 'Gd-ABIDJAN', '0'], dtype=object)

In [14]:
geo.replace('"""Abidjan_EST"', "Abidjan_EST",inplace=True)
geo.replace('"""Abidjan_CENTRE"', "Abidjan_CENTRE",inplace=True)
geo.replace('"""Abidjan_NORD"', "Abidjan_NORD",inplace=True)

In [15]:
geo["ZONENAME"].unique()

array(['AGBANNEW', '"ASSINIE"""', '"KRIKOREA"""', 'NPONON', 'OKAHOTEL',
       'PALMABINGERVILLE', 'TIAMA', 'WILLIAMSVILLE2',
       'ZONE E - Abidjan_NORD', 'ZONE B - Abidjan_EST',
       'ZONE D - Grand-OUEST', 'ZONE C - Grand-EST',
       'ZONE C - Abidjan_OUEST', 'ZONE B - Grand-SUD',
       'ZONE E - SUD-OUEST', 'ZONE A - Abidjan_CENTRE',
       'ZONE B - Abidjan_OUEST', 'ZONE B - Abidjan_CENTRE',
       'ZONE D - Abidjan_SUD', 'ZONE B - Grand-EST',
       'ZONE A - Grand-NORD', 'ZONE D - Grand-SUD',
       'ZONE D - Abidjan_EST', 'ZONE C - Grand-SUD',
       'ZONE C - Abidjan_SUD', 'ZONE A - Grand-SUD',
       'ZONE C - Abidjan_EST', 'ZONE B - Abidjan_Nord', '',
       'ZONE A - Grand-EST', 'ZONE B - Abidjan-EST',
       'ZONE A - Abidjan_OUEST', 'ZONE B - Abidjan_SUD',
       'ZONE B - Grand-NORD'], dtype=object)

In [16]:
geo.replace('"ASSINIE"""', "ASSINIE",inplace=True)
geo.replace('"KRIKOREA"""', "KRIKOREA",inplace=True)

In [17]:
geo["STATUS"].unique()

array(['In Service', ''], dtype=object)

In [18]:
geo["STATUS"].replace("","Not In Service", inplace=True)

In [19]:
geo["STATUS"].unique()

array(['In Service', 'Not In Service'], dtype=object)

In [20]:
geo["LOCALISATION"].unique()

array(['ABIDJAN', 'INTERIEUR'], dtype=object)

In [21]:
geo["REGION"].unique()

array(['LAGUNE', 'SUD COMOE', 'HAUT SASSANDRA', 'MOYEN COMOE', 'AGNEBY',
       'SUD BANDAMA', 'BAS SASSANDRA', 'LACS', 'FROMAGER',
       'VALLEE DU BANDAMA', 'NZI COMOE', 'ZANZAN', 'MOYEN CAVALLY',
       'WORODOUGOU', 'DENGUELE', '18 MONTAGNES', 'SAVANE', 'MARAHOUE',
       'BAFING'], dtype=object)

In [22]:
geo["AREA"].unique()

array(['', 'ABIDJAN NORTH', 'NORTH WEST', 'EAST', 'ABIDJAN SOUTH',
       'SOUTH WEST', 'CENTRE NORTH'], dtype=object)

In [23]:
geo["AREA"].replace('',np.nan,inplace=True)

In [24]:
geo["AREA"].unique()

array([nan, 'ABIDJAN NORTH', 'NORTH WEST', 'EAST', 'ABIDJAN SOUTH',
       'SOUTH WEST', 'CENTRE NORTH'], dtype=object)

In [25]:
geo["LONGITUDE"]=geo["LONGITUDE"].astype(float)
geo["LATITUDE"]=geo["LATITUDE"].astype(float)

In [26]:
geo.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3974 entries, 0 to 3973
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   VILLES        3974 non-null   object 
 1   STATUS        3974 non-null   object 
 2   LOCALISATION  3974 non-null   object 
 3   DECOUPZONE    3974 non-null   object 
 4   ZONENAME      3974 non-null   object 
 5   LONGITUDE     3974 non-null   float64
 6   LATITUDE      3974 non-null   float64
 7   REGION        3974 non-null   object 
 8   AREA          3951 non-null   object 
 9   CELL_ID       3974 non-null   object 
 10  SITE_CODE     3974 non-null   object 
dtypes: float64(2), object(9)
memory usage: 341.6+ KB


In [27]:
geo.isnull().sum()

VILLES           0
STATUS           0
LOCALISATION     0
DECOUPZONE       0
ZONENAME         0
LONGITUDE        0
LATITUDE         0
REGION           0
AREA            23
CELL_ID          0
SITE_CODE        0
dtype: int64

There are now 23 null values in the Area column

##### 2. Day 1

In [28]:
day1.describe(include="all")

PRODUTC        VALUE  ...     CELL_ID     SITE_ID
count     5001  5001.000000  ...        5001        4285
unique       3          NaN  ...        1790        1076
top      Voice          NaN  ...  ffa6759bb2  b47a998a3e
freq      2219          NaN  ...         716          64
mean       NaN    25.869626  ...         NaN         NaN
std        NaN   101.934502  ...         NaN         NaN
min        NaN     0.000000  ...         NaN         NaN
25%        NaN     0.000000  ...         NaN         NaN
50%        NaN     0.000000  ...         NaN         NaN
75%        NaN    25.000000  ...         NaN         NaN
max        NaN  4440.000000  ...         NaN         NaN

[11 rows x 10 columns]

- There are outliers in the value column. The mean is 25.869626	but the minimum value is 0 and the 75th percentile is 25 yet the maximum value is 4440

In [29]:
day1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5001 entries, 0 to 5000
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   PRODUTC          5001 non-null   object
 1   VALUE            5001 non-null   int64 
 2   DATETIME         5001 non-null   object
 3   CELL_ON_SITE     5001 non-null   object
 4   DW_A_NUMBER_INT  5001 non-null   object
 5   DW_B_NUMBER_INT  5001 non-null   object
 6   COUNTRY_A        5001 non-null   object
 7   COUNTRY_B        5001 non-null   object
 8   CELL_ID          5001 non-null   object
 9   SITE_ID          4285 non-null   object
dtypes: int64(1), object(9)
memory usage: 390.8+ KB


In [30]:
day1.isnull().sum()

PRODUTC              0
VALUE                0
DATETIME             0
CELL_ON_SITE         0
DW_A_NUMBER_INT      0
DW_B_NUMBER_INT      0
COUNTRY_A            0
COUNTRY_B            0
CELL_ID              0
SITE_ID            716
dtype: int64

- There are 716 null values in the site ID column


**Renaming column 1**

In [31]:
day1.rename(columns={"PRODUTC":"PRODUCT"},inplace=True)

In [32]:
day1.head()

PRODUCT  VALUE               DATETIME  ... COUNTRY_B     CELL_ID     SITE_ID
0   Voice      0  2012-05-06 23:04:37.0  ...     19e2e  /TJNe+Mmtu  /TJNe+Mmtu
1     sms      0  2012-05-06 23:05:25.0  ...     19e2e  /+cKZKIp41  /+cKZKIp41
2     sms     25  2012-05-06 23:05:41.0  ...     19e2e  /u0FSD+ahi  /u0FSD+ahi
3   Voice     56  2012-05-06 23:05:42.0  ...     19e2e  +cNeJzsTp3  +cNeJzsTp3
4   Voice     14  2012-05-06 23:06:29.0  ...     19e2e  051KhYDCpv  051KhYDCpv

[5 rows x 10 columns]

In [33]:
head=day1.head(10)
tail=day1.tail(10)
frames2=[head,tail]
head_tail=pd.concat(frames2)
head_tail

PRODUCT  VALUE               DATETIME  ... COUNTRY_B     CELL_ID     SITE_ID
0      Voice      0  2012-05-06 23:04:37.0  ...     19e2e  /TJNe+Mmtu  /TJNe+Mmtu
1        sms      0  2012-05-06 23:05:25.0  ...     19e2e  /+cKZKIp41  /+cKZKIp41
2        sms     25  2012-05-06 23:05:41.0  ...     19e2e  /u0FSD+ahi  /u0FSD+ahi
3      Voice     56  2012-05-06 23:05:42.0  ...     19e2e  +cNeJzsTp3  +cNeJzsTp3
4      Voice     14  2012-05-06 23:06:29.0  ...     19e2e  051KhYDCpv  051KhYDCpv
5       data      0  2012-05-06 23:07:26.0  ...     OTHER  /lrUXVGVcN  /lrUXVGVcN
6      Voice      0  2012-05-06 23:08:02.0  ...     19e2e  +38u2u/rfx  +38u2u/rfx
7      Voice      0  2012-05-06 23:08:03.0  ...     19e2e  09UjmmY1Ds  09UjmmY1Ds
8      Voice      0  2012-05-06 23:08:05.0  ...     19e2e  0DeimCpZOc  0DeimCpZOc
9        sms      0  2012-05-06 23:08:43.0  ...     19e2e  01MDiY5F89  01MDiY5F89
4991     sms      0  2012-05-07 00:01:54.0  ...     19e2e  85bbd0e6b0  481aaf0c3b
4992     sms     50  2012-05-07 00:01:54.0  ...     19e2e  6f3a2df445  f939f3349c
4993     sms     50  2012-05-07 00:01:54.0  ...     19e2e  babc0339bd  13c690c436
4994     sms      0  2012-05-07 00:01:54.0  ...     19e2e  0dc2b1dafc  ec013a6f3b
4995     sms      0  2012-05-07 00:01:54.0  ...     19e2e  5116d1c4a6  c07a544c63
4996     sms    100  2012-05-07 00:01:54.0  ...     OTHER  0c2371d9bc  1e751d08a4
4997   Voice      1  2012-05-07 00:01:54.0  ...     19e2e  5fb791cf56  65efe2bc62
4998   Voice      0  2012-05-07 00:01:54.0  ...     19e2e  7773627c12  9ef9616350
4999   Voice      0  2012-05-07 00:01:54.0  ...     19e2e  3f4af49d6c  969b8d0e39
5000   Voice      7  2012-05-07 00:01:54.0  ...     19e2e  6f4d9b77a5  ebfc58af70

[20 rows x 10 columns]

#### Changing the datetime column from string to datetime format, and introducing hour column

In [34]:
import datetime

In [35]:
day1_new=day1.copy()


In [36]:
day1_new["DATETIME"]=pd.to_datetime(day1_new["DATETIME"]).dt.to_period("d")

In [37]:
day1_new.head()

PRODUCT  VALUE    DATETIME  ... COUNTRY_B     CELL_ID     SITE_ID
0   Voice      0  2012-05-06  ...     19e2e  /TJNe+Mmtu  /TJNe+Mmtu
1     sms      0  2012-05-06  ...     19e2e  /+cKZKIp41  /+cKZKIp41
2     sms     25  2012-05-06  ...     19e2e  /u0FSD+ahi  /u0FSD+ahi
3   Voice     56  2012-05-06  ...     19e2e  +cNeJzsTp3  +cNeJzsTp3
4   Voice     14  2012-05-06  ...     19e2e  051KhYDCpv  051KhYDCpv

[5 rows x 10 columns]

In [38]:
day1_new.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5001 entries, 0 to 5000
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype    
---  ------           --------------  -----    
 0   PRODUCT          5001 non-null   object   
 1   VALUE            5001 non-null   int64    
 2   DATETIME         5001 non-null   period[D]
 3   CELL_ON_SITE     5001 non-null   object   
 4   DW_A_NUMBER_INT  5001 non-null   object   
 5   DW_B_NUMBER_INT  5001 non-null   object   
 6   COUNTRY_A        5001 non-null   object   
 7   COUNTRY_B        5001 non-null   object   
 8   CELL_ID          5001 non-null   object   
 9   SITE_ID          4285 non-null   object   
dtypes: int64(1), object(8), period[D](1)
memory usage: 390.8+ KB


In [39]:
def hr_function(ts):
    return ts.hour

In [40]:
day1["DATETIME"]=pd.to_datetime(day1["DATETIME"])

In [41]:
day1_new["Hour"]=day1["DATETIME"].apply(hr_function)

In [98]:
day1_new

PRODUCT  VALUE    DATETIME  ...     CELL_ID     SITE_ID Hour
0      Voice      0  2012-05-06  ...  /TJNe+Mmtu  /TJNe+Mmtu   23
1        sms      0  2012-05-06  ...  /+cKZKIp41  /+cKZKIp41   23
2        sms     25  2012-05-06  ...  /u0FSD+ahi  /u0FSD+ahi   23
3      Voice     56  2012-05-06  ...  +cNeJzsTp3  +cNeJzsTp3   23
4      Voice     14  2012-05-06  ...  051KhYDCpv  051KhYDCpv   23
...      ...    ...         ...  ...         ...         ...  ...
4996     sms    100  2012-05-07  ...  0c2371d9bc  1e751d08a4    0
4997   Voice      1  2012-05-07  ...  5fb791cf56  65efe2bc62    0
4998   Voice      0  2012-05-07  ...  7773627c12  9ef9616350    0
4999   Voice      0  2012-05-07  ...  3f4af49d6c  969b8d0e39    0
5000   Voice      7  2012-05-07  ...  6f4d9b77a5  ebfc58af70    0

[5001 rows x 11 columns]

In [43]:
day1_new["Hour"].unique()

array([23,  0])

##### Day 2

In [44]:
day2.describe(include="all")

PRODUCT        VALUE  ...     CELL_ID     SITE_ID
count     5001  5001.000000  ...        5001        4325
unique       3          NaN  ...        1661        1027
top        sms          NaN  ...  ffa6759bb2  91e86be9a6
freq      2329          NaN  ...         676          57
mean       NaN    24.334933  ...         NaN         NaN
std        NaN    97.471586  ...         NaN         NaN
min        NaN     0.000000  ...         NaN         NaN
25%        NaN     0.000000  ...         NaN         NaN
50%        NaN     0.000000  ...         NaN         NaN
75%        NaN    25.000000  ...         NaN         NaN
max        NaN  3380.000000  ...         NaN         NaN

[11 rows x 10 columns]

- There are outliers in the value column. The mean is 24.334933	but the minimum value is 0 and the 75th percentile is 25, yet the maximum value is 3380.

- There are null values in the site ID column

**Renaming columns**

In [45]:
day2.rename(columns={"DATE_TIME":"DATETIME"},inplace=True)
day2.rename(columns={"DW_A_NUMBER":"DW_A_NUMBER_INT"},inplace=True)
day2.rename(columns={"DW_B_NUMBER":"DW_B_NUMBER_INT"},inplace=True)

#### Changing the datetime column from string to datetime format, and introducing hour column

In [46]:
day2["DATETIME"]=pd.to_datetime(day2["DATETIME"])

In [47]:
day2_new=day2.copy()

In [48]:
day2_new["DATETIME"]=pd.to_datetime(day2_new["DATETIME"]).dt.to_period("d")

In [49]:
day2_new["Hour"]=day2["DATETIME"].apply(hr_function)

In [139]:
day2

PRODUCT  VALUE            DATETIME  ... COUNTRY_B     CELL_ID     SITE_ID
0        sms      0 2012-05-07 23:02:06  ...     19e2e  +854AcBQT2  +854AcBQT2
1      Voice      0 2012-05-07 23:03:44  ...     19e2e  +854AcBQT2  +854AcBQT2
2        sms      0 2012-05-07 23:04:06  ...     19e2e  +laSrk7g6q  +laSrk7g6q
3      Voice     37 2012-05-07 23:04:48  ...     19e2e  /xvRrCVKoQ  /xvRrCVKoQ
4        sms     25 2012-05-07 23:04:58  ...     19e2e  +bO+qkSonO  +bO+qkSonO
...      ...    ...                 ...  ...       ...         ...         ...
4996     sms      0 2012-05-08 00:01:53  ...     19e2e  896459aace  b7d00e59e4
4997     sms     25 2012-05-08 00:01:53  ...     19e2e  7e90b35166  211015b14d
4998     sms      0 2012-05-08 00:01:53  ...     19e2e  fe70e431df  c7ea989264
4999    data      0 2012-05-08 00:01:53  ...     OTHER  ffa6759bb2         NaN
5000   Voice      0 2012-05-08 00:01:53  ...     19e2e  8630058aca  b47a998a3e

[5001 rows x 10 columns]

In [121]:
day2_new.groupby(["PRODUCT","Hour"])["VALUE"].nlargest(10)

PRODUCT  Hour      
Voice    0     1071    3380
               1845    2330
               2780    2212
               1992    1412
               1707    1026
               3806    1007
               3864    1000
               3834     990
               4599     816
               2368     746
         23    457     1496
               570      400
               162      346
               406      307
               601      280
               138      231
               275      210
               105      200
               120      200
               505      200
data     0     3460    1000
               4923     676
               4199     464
               2212     371
               4620     330
               3380     209
               2055     205
               2381     152
               3342     148
               1556     126
         23    445      748
               573      190
               205       20
               336       10
               757        6


##### Day 3

In [50]:
day3.describe(include="all")

PRODUCT         VALUE  ...      CELLID     SIET_ID
count     5001   5001.000000  ...        5001        4396
unique       3           NaN  ...        1596         961
top        sms           NaN  ...  ffa6759bb2  1ddea341cd
freq      2767           NaN  ...         604          48
mean       NaN     26.441512  ...         NaN         NaN
std        NaN    224.200217  ...         NaN         NaN
min        NaN      0.000000  ...         NaN         NaN
25%        NaN      0.000000  ...         NaN         NaN
50%        NaN      0.000000  ...         NaN         NaN
75%        NaN     25.000000  ...         NaN         NaN
max        NaN  12900.000000  ...         NaN         NaN

[11 rows x 10 columns]

- There are outliers in the value column. The mean is 26.441512	but the minimum value is 0 and the 75th percentile is 25, yet the maximum value is 12900.

- There are null values in the site ID column


**Renaming the columns**

In [51]:
day3.rename(columns={"SIET_ID":"SITE_ID"},inplace=True)
day3.rename(columns={"DATE_TIME":"DATETIME"},inplace=True)
day3.rename(columns={"CELLID":"CELL_ID"},inplace=True)

#### Changing the datetime column from string to datetime format, and introducing hour column

In [52]:
day3["DATETIME"]=pd.to_datetime(day3["DATETIME"])

In [53]:
day3_new=day3.copy()

In [54]:
day3_new["DATETIME"]=pd.to_datetime(day3_new["DATETIME"]).dt.to_period("d")

In [55]:
day3_new["Hour"]=day3["DATETIME"].apply(hr_function)

In [103]:
day3_new

PRODUCT  VALUE    DATETIME  ...     CELL_ID     SITE_ID Hour
0      Voice     61  2012-05-08  ...  +38u2u/rfx  +38u2u/rfx   23
1        sms      0  2012-05-08  ...  +Fs4mTvdKx  +Fs4mTvdKx   23
2      Voice     10  2012-05-08  ...  +ZaaZozy/+  +ZaaZozy/+   23
3        sms      0  2012-05-08  ...  +cNeJzsTp3  +cNeJzsTp3   23
4      Voice     22  2012-05-08  ...  /eeMklgyJA  /eeMklgyJA   23
...      ...    ...         ...  ...         ...         ...  ...
4996   Voice      0  2012-05-09  ...  2aa04dcab7  f7d63a29e5    0
4997   Voice     62  2012-05-09  ...  7f1f2f889f  b4ddb9e93c    0
4998   Voice     32  2012-05-09  ...  a4731e780e  d68260ef50    0
4999   Voice     30  2012-05-09  ...  98ed726da2  b4130b54ab    0
5000   Voice     60  2012-05-09  ...  1c11a2fcbb  6d70dae094    0

[5001 rows x 11 columns]

In [138]:
day3

PRODUCT  VALUE            DATETIME  ... COUNTRY_B     CELL_ID     SITE_ID
0      Voice     61 2012-05-08 23:01:28  ...     19e2e  +38u2u/rfx  +38u2u/rfx
1        sms      0 2012-05-08 23:01:31  ...     19e2e  +Fs4mTvdKx  +Fs4mTvdKx
2      Voice     10 2012-05-08 23:02:59  ...     19e2e  +ZaaZozy/+  +ZaaZozy/+
3        sms      0 2012-05-08 23:03:04  ...     19e2e  +cNeJzsTp3  +cNeJzsTp3
4      Voice     22 2012-05-08 23:03:16  ...     19e2e  /eeMklgyJA  /eeMklgyJA
...      ...    ...                 ...  ...       ...         ...         ...
4996   Voice      0 2012-05-09 00:01:43  ...     19e2e  2aa04dcab7  f7d63a29e5
4997   Voice     62 2012-05-09 00:01:43  ...     19e2e  7f1f2f889f  b4ddb9e93c
4998   Voice     32 2012-05-09 00:01:43  ...     19e2e  a4731e780e  d68260ef50
4999   Voice     30 2012-05-09 00:01:43  ...     19e2e  98ed726da2  b4130b54ab
5000   Voice     60 2012-05-09 00:01:43  ...     19e2e  1c11a2fcbb  6d70dae094

[5001 rows x 10 columns]

#### More exploration

**_Average billing price per day per product_**

In [102]:
day1_new.groupby("Hour")["VALUE"].mean()

Hour
0     26.863102
23    19.358006
Name: VALUE, dtype: float64

In [57]:
day1.groupby("PRODUCT")["VALUE"].mean()

PRODUCT
Voice    49.062190
data      8.713755
sms       6.821772
Name: VALUE, dtype: float64

In [58]:
day2_new.groupby("Hour")["VALUE"].mean()

Hour
0     25.949941
23    15.406005
Name: VALUE, dtype: float64

In [59]:
day2.groupby("PRODUCT")["VALUE"].mean()

PRODUCT
Voice    54.294858
data      7.180157
sms       5.458566
Name: VALUE, dtype: float64

In [100]:
day3_new.groupby("Hour")["VALUE"].mean()

Hour
0     26.872109
23    23.228426
Name: VALUE, dtype: float64

In [60]:
day3.groupby("PRODUCT")["VALUE"].mean()

PRODUCT
Voice    65.047372
data      6.861472
sms       9.844958
Name: VALUE, dtype: float64

**_Total billing price per day per product_**

In [61]:
day1.groupby("PRODUCT")["VALUE"].sum()

PRODUCT
Voice    108869
data       7032
sms       13473
Name: VALUE, dtype: int64

In [62]:
day2.groupby("PRODUCT")["VALUE"].sum()

PRODUCT
Voice    103486
data       5500
sms       12713
Name: VALUE, dtype: int64

In [63]:
day3.groupby("PRODUCT")["VALUE"].sum()

PRODUCT
Voice    100238
data       4755
sms       27241
Name: VALUE, dtype: int64

#### Joining the dataframes

In [64]:
frames=[day1_new,day2_new,day3_new]
all_days=pd.concat(frames,ignore_index=True)

In [65]:
all_days.head()

PRODUCT  VALUE    DATETIME  ...     CELL_ID     SITE_ID Hour
0   Voice      0  2012-05-06  ...  /TJNe+Mmtu  /TJNe+Mmtu   23
1     sms      0  2012-05-06  ...  /+cKZKIp41  /+cKZKIp41   23
2     sms     25  2012-05-06  ...  /u0FSD+ahi  /u0FSD+ahi   23
3   Voice     56  2012-05-06  ...  +cNeJzsTp3  +cNeJzsTp3   23
4   Voice     14  2012-05-06  ...  051KhYDCpv  051KhYDCpv   23

[5 rows x 11 columns]

In [66]:
all_days.tail()

PRODUCT  VALUE    DATETIME  ...     CELL_ID     SITE_ID Hour
14998   Voice      0  2012-05-09  ...  2aa04dcab7  f7d63a29e5    0
14999   Voice     62  2012-05-09  ...  7f1f2f889f  b4ddb9e93c    0
15000   Voice     32  2012-05-09  ...  a4731e780e  d68260ef50    0
15001   Voice     30  2012-05-09  ...  98ed726da2  b4130b54ab    0
15002   Voice     60  2012-05-09  ...  1c11a2fcbb  6d70dae094    0

[5 rows x 11 columns]

In [67]:
all_days.shape

(15003, 11)

##### Average billing price per product over the three days

In [68]:
all_days.groupby("PRODUCT")["VALUE"].mean()

PRODUCT
Voice    55.169961
data      7.628861
sms       7.555791
Name: VALUE, dtype: float64

##### Total billing price per product over the three days

In [69]:
all_days.groupby("PRODUCT")["VALUE"].sum()

PRODUCT
Voice    312593
data      17287
sms       53427
Name: VALUE, dtype: int64

### Inner and left join the geo  dataframe

In [70]:
geo.rename(columns={"SITE_CODE":"SITE_ID"},inplace=True)

**_Inner join by the site ID column gets rid of the null SITE ID values_**

In [71]:
CDR_description

Column name                                        Description   Format
0          PRODUCT                                       Voice or SMS   String
1            VALUE                                      Billing price  Integer
2        DATE_TIME               Time in format yyyy-MM-dd hh:mm:ss.0   String
3     CELL_ON_SITE  Which cell in the site was used (not needed here)  Integer
4  DW_A_NUMBER_INT  Anonymized phone number of the person for whic...   String
5  DW_B_NUMBER_INT        Anonymized phone number of the counterparty   String
6        COUNTRY_A                  Country of party A (useless here)   String
7        COUNTRY_B                  Country of party B (useless here)   String
8          CELL_ID                                     ID of the cell   String
9          SITE_ID                                     ID of the SITE   String

In [72]:
geo2=geo.drop("CELL_ID",axis=1)

In [73]:
geo2.head()

VILLES      STATUS LOCALISATION  ...     REGION AREA     SITE_ID
0   ADJAME  In Service      ABIDJAN  ...     LAGUNE  NaN  0bb492ef17
1   ADJAME  In Service      ABIDJAN  ...     LAGUNE  NaN  0bb492ef17
2   ADJAME  In Service      ABIDJAN  ...     LAGUNE  NaN  0bb492ef17
3  ASSINIE  In Service    INTERIEUR  ...  SUD COMOE  NaN  8a1e6ce31f
4  ASSINIE  In Service    INTERIEUR  ...  SUD COMOE  NaN  8a1e6ce31f

[5 rows x 10 columns]

In [74]:
all_days_2=all_days.drop("CELL_ID",axis=1)

In [75]:
joined_inner=pd.merge(all_days_2,
                      geo2,
                     how="inner",
                     on="SITE_ID")

In [76]:
head=joined_inner.head(10)
tail=joined_inner.tail(10)
frames2=[head,tail]
head_tail=pd.concat(frames2)
head_tail

PRODUCT  VALUE    DATETIME  ...  LATITUDE             REGION           AREA
0         sms      0  2012-05-07  ...  7.718417  VALLEE DU BANDAMA   CENTRE NORTH
1         sms      0  2012-05-07  ...  7.718417  VALLEE DU BANDAMA   CENTRE NORTH
2         sms      0  2012-05-07  ...  7.718417  VALLEE DU BANDAMA   CENTRE NORTH
3         sms      0  2012-05-07  ...  7.718417  VALLEE DU BANDAMA   CENTRE NORTH
4         sms      0  2012-05-07  ...  7.718417  VALLEE DU BANDAMA   CENTRE NORTH
5         sms      0  2012-05-07  ...  7.718417  VALLEE DU BANDAMA   CENTRE NORTH
6       Voice     92  2012-05-07  ...  7.718417  VALLEE DU BANDAMA   CENTRE NORTH
7       Voice     92  2012-05-07  ...  7.718417  VALLEE DU BANDAMA   CENTRE NORTH
8       Voice     92  2012-05-07  ...  7.718417  VALLEE DU BANDAMA   CENTRE NORTH
9       Voice     92  2012-05-07  ...  7.718417  VALLEE DU BANDAMA   CENTRE NORTH
51351     sms      0  2012-05-09  ...  5.138750          SUD COMOE  ABIDJAN SOUTH
51352     sms      0  2012-05-09  ...  5.138750          SUD COMOE  ABIDJAN SOUTH
51353   Voice     47  2012-05-09  ...  6.843278           MARAHOUE   CENTRE NORTH
51354   Voice     47  2012-05-09  ...  6.843278           MARAHOUE   CENTRE NORTH
51355   Voice     47  2012-05-09  ...  6.843278           MARAHOUE   CENTRE NORTH
51356   Voice     17  2012-05-09  ...  5.349442             LAGUNE  ABIDJAN NORTH
51357   Voice     17  2012-05-09  ...  5.349442             LAGUNE  ABIDJAN NORTH
51358     sms      0  2012-05-09  ...  7.777056  VALLEE DU BANDAMA   CENTRE NORTH
51359     sms      0  2012-05-09  ...  7.777056  VALLEE DU BANDAMA   CENTRE NORTH
51360     sms      0  2012-05-09  ...  7.777056  VALLEE DU BANDAMA   CENTRE NORTH

[20 rows x 19 columns]

In [77]:
joined_inner.drop_duplicates(inplace=True)

In [78]:
head=joined_inner.head(10)
tail=joined_inner.tail(10)
frames2=[head,tail]
head_tail=pd.concat(frames2)
head_tail

PRODUCT  VALUE    DATETIME  ...  LATITUDE             REGION           AREA
0         sms      0  2012-05-07  ...  7.718417  VALLEE DU BANDAMA   CENTRE NORTH
6       Voice     92  2012-05-07  ...  7.718417  VALLEE DU BANDAMA   CENTRE NORTH
12        sms     25  2012-05-07  ...  7.718417  VALLEE DU BANDAMA   CENTRE NORTH
18      Voice     16  2012-05-07  ...  7.718417  VALLEE DU BANDAMA   CENTRE NORTH
24      Voice     35  2012-05-07  ...  7.718417  VALLEE DU BANDAMA   CENTRE NORTH
30        sms     25  2012-05-07  ...  7.718417  VALLEE DU BANDAMA   CENTRE NORTH
36      Voice     19  2012-05-07  ...  7.718417  VALLEE DU BANDAMA   CENTRE NORTH
42      Voice     20  2012-05-07  ...  7.718417  VALLEE DU BANDAMA   CENTRE NORTH
48      Voice     60  2012-05-07  ...  7.718417  VALLEE DU BANDAMA   CENTRE NORTH
54      Voice     46  2012-05-07  ...  7.718417  VALLEE DU BANDAMA   CENTRE NORTH
51328     sms      0  2012-05-09  ...  5.580444        SUD BANDAMA     SOUTH WEST
51334   Voice      5  2012-05-09  ...  6.980417          NZI COMOE   CENTRE NORTH
51337   Voice     68  2012-05-09  ...  6.914972          NZI COMOE           EAST
51341   Voice     22  2012-05-09  ...  4.931389      BAS SASSANDRA     SOUTH WEST
51344     sms     50  2012-05-09  ...  5.553722             LAGUNE   CENTRE NORTH
51347   Voice      0  2012-05-09  ...  7.170250           MARAHOUE   CENTRE NORTH
51350     sms      0  2012-05-09  ...  5.138750          SUD COMOE  ABIDJAN SOUTH
51353   Voice     47  2012-05-09  ...  6.843278           MARAHOUE   CENTRE NORTH
51356   Voice     17  2012-05-09  ...  5.349442             LAGUNE  ABIDJAN NORTH
51358     sms      0  2012-05-09  ...  7.777056  VALLEE DU BANDAMA   CENTRE NORTH

[20 rows x 19 columns]

In [79]:
joined_inner.shape

(8856, 19)

In [80]:
joined_inner.reset_index(drop=True,inplace=True)

In [81]:
head=joined_inner.head(10)
tail=joined_inner.tail(10)
frames2=[head,tail]
head_tail=pd.concat(frames2)
head_tail

PRODUCT  VALUE    DATETIME  ...  LATITUDE             REGION           AREA
0        sms      0  2012-05-07  ...  7.718417  VALLEE DU BANDAMA   CENTRE NORTH
1      Voice     92  2012-05-07  ...  7.718417  VALLEE DU BANDAMA   CENTRE NORTH
2        sms     25  2012-05-07  ...  7.718417  VALLEE DU BANDAMA   CENTRE NORTH
3      Voice     16  2012-05-07  ...  7.718417  VALLEE DU BANDAMA   CENTRE NORTH
4      Voice     35  2012-05-07  ...  7.718417  VALLEE DU BANDAMA   CENTRE NORTH
5        sms     25  2012-05-07  ...  7.718417  VALLEE DU BANDAMA   CENTRE NORTH
6      Voice     19  2012-05-07  ...  7.718417  VALLEE DU BANDAMA   CENTRE NORTH
7      Voice     20  2012-05-07  ...  7.718417  VALLEE DU BANDAMA   CENTRE NORTH
8      Voice     60  2012-05-07  ...  7.718417  VALLEE DU BANDAMA   CENTRE NORTH
9      Voice     46  2012-05-07  ...  7.718417  VALLEE DU BANDAMA   CENTRE NORTH
8846     sms      0  2012-05-09  ...  5.580444        SUD BANDAMA     SOUTH WEST
8847   Voice      5  2012-05-09  ...  6.980417          NZI COMOE   CENTRE NORTH
8848   Voice     68  2012-05-09  ...  6.914972          NZI COMOE           EAST
8849   Voice     22  2012-05-09  ...  4.931389      BAS SASSANDRA     SOUTH WEST
8850     sms     50  2012-05-09  ...  5.553722             LAGUNE   CENTRE NORTH
8851   Voice      0  2012-05-09  ...  7.170250           MARAHOUE   CENTRE NORTH
8852     sms      0  2012-05-09  ...  5.138750          SUD COMOE  ABIDJAN SOUTH
8853   Voice     47  2012-05-09  ...  6.843278           MARAHOUE   CENTRE NORTH
8854   Voice     17  2012-05-09  ...  5.349442             LAGUNE  ABIDJAN NORTH
8855     sms      0  2012-05-09  ...  7.777056  VALLEE DU BANDAMA   CENTRE NORTH

[20 rows x 19 columns]

**Most used cities for the three days**

In [82]:
joined_inner.groupby(['VILLES'])["VALUE"].sum().nlargest(5)

VILLES
YOPOUGON    39837
COCODY      36023
ABOBO       23260
KOUMASSI    13647
ADJAME      12933
Name: VALUE, dtype: int64

**Most used cities for each of the three days**

In [140]:
joined_inner.groupby(['DATETIME','VILLES'])["VALUE"].sum().nlargest(10)

DATETIME    VILLES  
2012-05-09  COCODY      16478
2012-05-07  YOPOUGON    14518
2012-05-09  YOPOUGON    13679
2012-05-08  YOPOUGON    11640
2012-05-07  COCODY      10787
2012-05-08  COCODY       8758
2012-05-07  ABOBO        8099
2012-05-08  ABOBO        7609
2012-05-09  ABOBO        7552
2012-05-08  KOUMASSI     5232
Name: VALUE, dtype: int64

**Product use for the three days**

In [84]:
joined_inner.groupby(['PRODUCT','VILLES'])["VALUE"].sum()

PRODUCT  VILLES       
Voice    ABENGOUROU        1710
         ABOBO            18626
         ABOISSO            268
         ABOISSO COMOE      240
         ABONGOUA             1
                          ...  
sms      ZAIBO               50
         ZAROKO              25
         ZIKISSO            100
         ZOUKOUGBEU          50
         ZUENOULA           100
Name: VALUE, Length: 601, dtype: int64

In [94]:
joined_inner.groupby(['PRODUCT','VILLES'])["VALUE"].sum().nlargest()

PRODUCT  VILLES     
Voice    COCODY         32723
         YOPOUGON       32299
         ABOBO          18626
         TREICHVILLE    11748
         KOUMASSI       11472
Name: VALUE, dtype: int64

In [137]:
joined_inner.groupby(["PRODUCT","VILLES","Hour","DATETIME"])["VALUE"].sum().nlargest(50)

PRODUCT  VILLES        Hour  DATETIME  
Voice    COCODY        0     2012-05-09    14878
         YOPOUGON      0     2012-05-07    12718
                             2012-05-08    10407
         COCODY        0     2012-05-07     9837
         YOPOUGON      0     2012-05-09     9174
         COCODY        0     2012-05-08     8008
         ABOBO         0     2012-05-08     7109
                             2012-05-07     6974
         TREICHVILLE   0     2012-05-09     4559
         ABOBO         0     2012-05-09     4543
sms      YOPOUGON      0     2012-05-09     4505
Voice    KOUMASSI      0     2012-05-08     4457
         ADJAME        0     2012-05-08     4221
                             2012-05-07     4156
         KOUMASSI      0     2012-05-09     3922
         TREICHVILLE   0     2012-05-07     3605
                             2012-05-08     3584
         MARCORY       0     2012-05-08     3255
                             2012-05-07     3169
         SAN-PEDRO     0     

In [143]:
joined_inner.groupby(["PRODUCT","VILLES","Hour","DATETIME"])["VALUE"].sum().nsmallest(50)

PRODUCT  VILLES             Hour  DATETIME  
Voice    ABOISSO            0     2012-05-08    0
         ABOUDEKOUASSIKRO   0     2012-05-07    0
         AGONDA             0     2012-05-08    0
         AKOUYAKRO          0     2012-05-07    0
         ALEPE              0     2012-05-07    0
         AZAGUIE            0     2012-05-08    0
         BALEKO             0     2012-05-07    0
         BANGOLO            0     2012-05-08    0
         BAZRA NATTIS       0     2012-05-07    0
         BEOUMI             0     2012-05-09    0
         BOGUEDIA           0     2012-05-08    0
         BONON              0     2012-05-09    0
         BONOUA             0     2012-05-08    0
         BONOUFLA           0     2012-05-07    0
         BOUADIKRO          0     2012-05-07    0
         BOUNDIALI          0     2012-05-08    0
         BROBO              0     2012-05-07    0
         COLLINEKASSEREGUE  0     2012-05-07    0
         DAGADJI            0     2012-05-08    0
     

# **Uploading to github**

In [144]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [145]:
cd '/content/drive/My Drive/GEOFFREY CHEGE MWANGI I.P. W3'

/content/drive/My Drive/GEOFFREY CHEGE MWANGI I.P. W3


In [146]:
username = 'Geoffrey-Chege'
repository = 'Moringa-School-IP-W3-Geoffrey-Chege'
git_token = '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'

In [149]:
!git init Moringa-School-IP-W3-Geoffrey-Chege

Initialized empty Git repository in /content/drive/My Drive/GEOFFREY CHEGE MWANGI I.P. W3/Moringa-School-IP-W3-Geoffrey-Chege/.git/


In [150]:
cd Moringa-School-IP-W3-Geoffrey-Chege/

/content/drive/My Drive/GEOFFREY CHEGE MWANGI I.P. W3/Moringa-School-IP-W3-Geoffrey-Chege


In [160]:
!git status

On branch master

No commits yet

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	Copy of Copy of Moringa_Data_Science_Prep_W3_Independent_Project_2021_09_Geoffrey_Mwangi_DataReport
	Copy of Moringa_Data_Science_Prep_W3_Independent_Project_2021_09_GEOFFREY_CHEGE_DataReport.gdoc

nothing added to commit but untracked files present (use "git add" to track)


In [161]:
!git config --global user.name "Geoffrey-Chege"
!git config --global user.email "geoffrey.chege@student.moringaschool.com"

In [162]:
!git add .

error: open("Copy of Moringa_Data_Science_Prep_W3_Independent_Project_2021_09_GEOFFREY_CHEGE_DataReport.gdoc"): Operation not supported
error: unable to index file Copy of Moringa_Data_Science_Prep_W3_Independent_Project_2021_09_GEOFFREY_CHEGE_DataReport.gdoc
fatal: adding files failed


In [156]:
#!echo "# Moringa-School-IP-W3-Geoffrey-Chege"
!git commit -m "second commit"
!git branch -M main
!git remote add origin https://{git_token}@github.com/{username}/{repository}.git
!git push -u origin main

#!git remote add origin https://github.com/Geoffrey-Chege/Moringa-School-IP-W3-Geoffrey-Chege.git
#!git branch -M main
#!git push -u origin main

error: open("Copy of Moringa_Data_Science_Prep_W3_Independent_Project_2021_09_GEOFFREY_CHEGE_DataReport.gdoc"): Operation not supported
error: unable to index file Copy of Moringa_Data_Science_Prep_W3_Independent_Project_2021_09_GEOFFREY_CHEGE_DataReport.gdoc
fatal: adding files failed
On branch master

Initial commit

Untracked files:
	Copy of Copy of Moringa_Data_Science_Prep_W3_Independent_Project_2021_09_Geoffrey_Mwangi_DataReport
	Copy of Moringa_Data_Science_Prep_W3_Independent_Project_2021_09_GEOFFREY_CHEGE_DataReport.gdoc

nothing added to commit but untracked files present
error: refname refs/heads/master not found
fatal: Branch rename failed
fatal: remote origin already exists.
error: src refspec main does not match any.
error: failed to push some refs to 'https://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@github.com/Geoffrey-Chege/Moringa-School-IP-W3-Geoffrey-Chege.git'
